lib/ActivityStreams.js

```diff
@@ -0,0 +1,144 @@
+const {PlacesProvider} = require("lib/PlacesProvider");
+
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource:///modules/NewTabURL.jsm");
+
+XPCOMUtils.defineLazyGetter(this, "EventEmitter", function () {
+  const {EventEmitter} = Cu.import("resource://devtools/shared/event-emitter.js", {});
+  return EventEmitter;
+});
+
+let ActivityStreams = function (options) {
+  console.debug(`ActivityStreams.init`);
+
+  this.options = options ? options : {};
+  this.pageURL = this.options.pageURL ? options.pageURL : data.url("content/activity-streams.html");
+
+  this.workers = new Set();
+  EventEmitter.decorate(this);
+
+  this._setupPageMod();
+  this._setupListeners();
+  NewTabURL.override(this.pageURL);
+};
+
+ActivityStreams.prototype = {
+
+  pageMod: null,
+
+  /**
+   * Send a message to a worker
+   */
+  send(actionName, msg, worker) {
+    worker.port.emit("addon-to-content", {
+      type: actionName,
+      data: msg
+    });
+  },
+
+  /**
+   * Broadcast a message to all workers
+   */
+  broadcast(actionName, msg) {
+    for (let worker of this.workers) {
+      this.send(actionName, msg, worker);
+    }
+  },
+
+  /**
+   * Responds to places requests
+   */
+  _respondToPlacesRequests(msgName, params) {
+    let {msg, worker} = params;
+    PlacesProvider.links.getTopFrecentSites(msg.data).then(links => {
+      this.send("RECEIVE_TOP_FRECENT_SITES", links, worker);
+    });
+  },
+
+
+  /**
+   * Broadcast places changes to pages
+   */
+  _handlePlacesChanges(eventName, data) {
+    this.broadcast("RECEIVE_PLACES_CHANGES", {
+      "event": eventName,
+      "data": data
+    });
+  },
+
+  /**
+   * Sets up various listeners for the pages
+   */
+  _setupListeners() {
+    PlacesProvider.links.on("deleteURI", this._handlePlacesChanges.bind(this));
+    PlacesProvider.links.on("clearHistory", this._handlePlacesChanges.bind(this));
+    PlacesProvider.links.on("linkChanged", this._handlePlacesChanges.bind(this));
+    PlacesProvider.links.on("manyLinksChanged", this._handlePlacesChanges.bind(this));
+
+    this.on("REQUEST_TOP_FRECENT_SITES", this._respondToPlacesRequests.bind(this));
+  },
+
+  /**
+   * Turns off various listeners for the pages
+   */
+  _removeListeners() {
+    PlacesProvider.links.off("deleteURI", this._handlePlacesChanges);
+    PlacesProvider.links.off("clearHistory", this._handlePlacesChanges);
+    PlacesProvider.links.off("linkChanged", this._handlePlacesChanges);
+    PlacesProvider.links.off("manyLinksChanged", this._handlePlacesChanges);
+
+    this.off("REQUEST_TOP_FRECENT_SITES", this._respondToPlacesRequests);
+  },
+
+  /**
+   * Sets up communications with the pages
+   */
+  _setupPageMod() {
+    this.pagemod = new PageMod({
+      include: [this.pageURL],
+      contentScriptFile: data.url("content-bridge.js"),
+      contentScriptWhen: "start",
+      attachTo: "top",
+      onAttach: worker => {
+        if (!this.workers.has(worker)) {
```

Comment:
I honestly don't know from reading the code :crying_cat_face:, which is why I asked. It would be even more confusing if it's the prototype... as that adds a 4th potential of ambiguity of `this`... JS, I :heart: you, I :broken_heart: you.